# The Voluntary Classroom Questionnaire

The Voluntary Classroom Questionnaire (VCQ) dataset was developed to offer students a manageable and user-friendly resource for statistical analysis. It consists of 154 variables, including demographic questions and multiple-item measures in areas such as school, work, friends and family, and well-being. The data was collected from 2015 to 2019 in Dr. Trinidad’s Introduction to Statistics for the Social Sciences (SOC 201) and Quantitative Research Methods (SOC 360) classes. Approximately 543 students participated in the VCQ over the five-year collection period.

The VCQ was administered as an in-class activity for educational purposes. Students were invited to complete a 23-page paper-pencil survey, but submission was not mandatory. They were informed that they could skip any question they found uncomfortable and could stop at any time. To ensure anonymity, students were told not to include their names and to skip or fabricate values for any demographic variables that might be outliers.
Students who chose to submit their completed questionnaire did so after Dr. Trinidad left the room. They were then instructed to randomly mix their submitted questionnaires with those from other classes and previous semesters.

Participation in the VCQ offered several educational benefits to students: (1) It provided students with firsthand experience of participating in a survey, (2) It helped students understand the processes of data entry and data cleaning, as they were required to practice both, and (3) It provided students with a dataset for practicing statistical analysis.
It’s important to note that the VCQ was not designed for publication analysis, as students fabricated some of their responses to demographic questions to maintain anonymity. The primary objective of the VCQ was to provide students with a hands-on learning experience. The VCQ continues to be a valuable educational tool, offering students firsthand experience in data collection, entry, cleaning, and analysis.

Let's first review the [VCQ code book](https://docs.google.com/document/d/1uYfUuUsPVQyD3puVvbIMjamYuVPr_f4gtJuSB5Wj-WA/edit) prepared by Dr. Trinidad.

## Practice

Write down some questions you have about the dataset.

Write down some questions about students who completed the survey you think could be answered by the dataset.

# Data Preparation

The raw data was provided by Dr. Trinidad in the Comma Separated Values (CSV) format. The CSV file has been uploaded to GitHub: [Raw data](https://raw.githubusercontent.com/qarnac/cs201/main/vcq_data_nan.csv). The following code block reads the CSV file into a DataFrame object for further exploration. In addition, it calls the shape function to find out the numbers of rows and columns in the dataset.

In [ ]:
#Collect data into a DataFrame object
import pandas as pd
raw_df = pd.read_csv('https://raw.githubusercontent.com/qarnac/cs201/main/vcq_data_nan.csv')
raw_df.shape

(543, 154)

The above result shows that there are 543 rows and 154 columns.

In the past, we learned a lot about the columns of a dataset using the `info` function.

In [ ]:
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 543 entries, 0 to 542
Columns: 154 entries, id to satisfaction_performancework
dtypes: float64(134), int64(20)
memory usage: 653.4 KB


However, since there are so many columns in the dataset, the above result only provides a summary. Out of the 154 columns, 134 are of float64 data type and 20 are of int64 data type.

Next, let's call the `head` function to see some example data from the dataset.

In [ ]:
raw_df.head()

,id,demo_gender,demo_age,demo_relationshipstatus01,demo_children,demo_classstanding,demo_units,demo_gpa,demo_employed01,demo_employed02,...,coworkers_respect,coworkers_friends,coworkers_argue,coworkers_like,wis_time,wis_tired,satisfaction_performanceschool,siw_time,siw_tired,satisfaction_performancework
0,1,1,19.0,1,2.0,2,15.0,3.15,1.0,1,...,4.0,5.0,5.0,5.0,4.0,4.0,3.0,3.0,3.0,2.0
1,2,2,26.0,2,2.0,3,14.0,3.30,1.0,1,...,6.0,6.0,1.0,6.0,5.0,5.0,4.0,5.0,5.0,4.0
2,3,1,NaN,1,2.0,3,15.0,3.11,2.0,2,...,NaN,NaN,NaN,NaN,3.0,2.0,NaN,2.0,3.0,3.0
3,4,2,NaN,1,2.0,2,17.0,3.40,2.0,2,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,5,2,19.0,1,2.0,2,13.0,3.00,2.0,2,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Since there are so many columns, not all columns are listed. We will first look at the columns for the demographics data.

# The filter function

In pandas library, the filter() function provides a flexible way to select columns based on patterns in their names. There are three common parameters we could use with the function.

*  The items parameter is used when we choose columns using their exact names.
*  The like parameter is used when we select columns that contain a substring.
*  The regex parameter supports more flexible matching such as start with or end with a substring.

This makes filter() especially useful when column names follow consistent naming conventions.

In [ ]:
# The items parameter takes a list of column names
gender_age_df = raw_df.filter(items=['demo_gender', 'demo_age']).copy()
gender_age_df.head()

In [ ]:
# This like parameter looks for all columns whose names include 'employ'
employ_df = raw_df.filter(like='employ').copy()
employ_df.columns

In [ ]:
# This regex parameter looks for all columns whose names start with demo
demo_df = raw_df.filter(regex='^demo').copy()
demo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 543 entries, 0 to 542
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   demo_gender                543 non-null    int64  
 1   demo_age                   534 non-null    float64
 2   demo_relationshipstatus01  543 non-null    int64  
 3   demo_children              542 non-null    float64
 4   demo_classstanding         543 non-null    int64  
 5   demo_units                 539 non-null    float64
 6   demo_gpa                   518 non-null    float64
 7   demo_employed01            542 non-null    float64
 8   demo_employed02            543 non-null    int64  
 9   demo_employed03            388 non-null    float64
 10  demo_incomehousehold       332 non-null    float64
 11  demo_incomepersonal        366 non-null    float64
 12  demo_politics              505 non-null    float64
 13  demo_religion              539 non-null    float64

## Practice

In [ ]:
# Use the items parameter to create a new DataFrame object that contain three columns of your choice
# Be sure the call the copy function so that any change to the new DataFrame will not affect raw_df
stress_df = raw_df.filter(items=['stress_relax', 'stress_concentrate', 'stress_sleep']).copy()

,stress_relax,stress_concentrate,stress_sleep
0,5.0,5.0,6.0
1,6.0,6.0,6.0
2,3.0,5.0,5.0
3,6.0,6.0,6.0
4,4.0,3.0,3.0
...,...,...,...
538,6.0,6.0,6.0
539,6.0,6.0,6.0
540,4.0,4.0,2.0
541,4.0,2.0,2.0


In [ ]:
# Use the like parameter to create a new DataFrame object that contains columns whose names include relation
# Be sure the call the copy function so that any change to the new DataFrame will not affect raw_df
raw_df.filter(like='relation')

,demo_relationshipstatus01,demo_relationshipstatus02
0,1,1.0
1,2,2.0
2,1,1.0
3,1,1.0
4,1,1.0
...,...,...
538,1,1.0
539,1,1.0
540,2,2.0
541,1,1.0


In [ ]:
# Use the regex parameter to create a new DataFrame object that contains columns whose names start with social
# Be sure the call the copy function so that any change to the new DataFrame will not affect raw_df
raw_df.filter(regex='^social')

,socialmedia_constantly,socialmedia_first,socialmedia_last,socialmedia_dumb
0,5.0,4.0,4.0,5.0
1,1.0,1.0,1.0,6.0
2,5.0,4.0,4.0,3.0
3,6.0,4.0,5.0,3.0
4,6.0,4.0,6.0,4.0
...,...,...,...,...
538,6.0,6.0,6.0,4.0
539,5.0,5.0,5.0,4.0
540,5.0,2.0,2.0,3.0
541,3.0,2.0,3.0,3.0


# The str accessor

Pandas provides a powerful set of string-handling tools through the .str accessor, which lets you apply string operations to a collection of text data at once. We will introduce more of them as we go but will first learn how to use `.str.removeprefix` to clean up column names.

In the following code block, we use `.str.removeprefix` with the columns of `demo_df`. This action will generate **a new list** of column names from the old one by removing the prefix `'demo_'` from them. It is important to assign the resulting new list back to `demo_df.columns`.

In [ ]:
demo_df.columns

Index(['demo_gender', 'demo_age', 'demo_relationshipstatus01', 'demo_children',
       'demo_classstanding', 'demo_units', 'demo_gpa', 'demo_employed01',
       'demo_employed02', 'demo_employed03', 'demo_incomehousehold',
       'demo_incomepersonal', 'demo_politics', 'demo_religion',
       'demo_relationshipstatus02', 'demo_employed04'],
      dtype='object')

In [ ]:
demo_df.columns = demo_df.columns.str.removeprefix('demo_')
demo_df.head()

,gender,age,relationshipstatus01,children,classstanding,units,gpa,employed01,employed02,employed03,incomehousehold,incomepersonal,politics,religion,relationshipstatus02,employed04
0,1,19.0,1,2.0,2,15.0,3.15,1.0,1,45.0,60000.0,16000.0,NaN,6.0,1.0,1.0
1,2,26.0,2,2.0,3,14.0,3.30,1.0,1,35.0,27000.0,27000.0,4.0,5.0,2.0,1.0
2,1,NaN,1,2.0,3,15.0,3.11,2.0,2,NaN,6000.0,2000.0,4.0,3.0,1.0,NaN
3,2,NaN,1,2.0,2,17.0,3.40,2.0,2,NaN,NaN,NaN,3.0,4.0,1.0,2.0
4,2,19.0,1,2.0,2,13.0,3.00,2.0,2,NaN,NaN,NaN,3.0,3.0,1.0,2.0


In [ ]:
raw_df.head()

,id,demo_gender,demo_age,demo_relationshipstatus01,demo_children,demo_classstanding,demo_units,demo_gpa,demo_employed01,demo_employed02,...,coworkers_respect,coworkers_friends,coworkers_argue,coworkers_like,wis_time,wis_tired,satisfaction_performanceschool,siw_time,siw_tired,satisfaction_performancework
0,1,1,19.0,1,2.0,2,15.0,3.15,1.0,1,...,4.0,5.0,5.0,5.0,4.0,4.0,3.0,3.0,3.0,2.0
1,2,2,26.0,2,2.0,3,14.0,3.30,1.0,1,...,6.0,6.0,1.0,6.0,5.0,5.0,4.0,5.0,5.0,4.0
2,3,1,NaN,1,2.0,3,15.0,3.11,2.0,2,...,NaN,NaN,NaN,NaN,3.0,2.0,NaN,2.0,3.0,3.0
3,4,2,NaN,1,2.0,2,17.0,3.40,2.0,2,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,5,2,19.0,1,2.0,2,13.0,3.00,2.0,2,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## Practice

In [ ]:
# Use the regex parameter with the filter function to create a new DataFrame object that contains
# columns whose names start with family. Be sure to call the copy function so that changes made
# to the new DataFrame will not affect raw_df.
family_df = raw_df.filter(regex='^family').copy()

#Use .str.removeprefix to remove the shared prefix
# from the column names in the new DataFrame object.
family_df.columns = family_df.columns.str.removeprefix('family_')

In [ ]:
family_df.head()

,enjoy,stress,sacrifices,like,secrets,fight,regularly
0,6.0,5.0,6.0,6.0,6.0,4.0,6.0
1,6.0,3.0,6.0,6.0,4.0,3.0,5.0
2,6.0,1.0,6.0,6.0,3.0,2.0,3.0
3,6.0,5.0,6.0,6.0,4.0,3.0,6.0
4,6.0,4.0,4.0,6.0,4.0,3.0,5.0


# Descriptive statistics

Looking at the code book, we can see that descriptive statistics would be more useful for some columns than the others. For example, although the gender column contains numeric data, they are used to represent categories: 1 for Male, 2 for Female, and 3 for Others. In such a case, it does not make much sense to determine the average or standard deviation for the column. The age column, on the other hand, has a winder range of numeric values. It would be interesting to see some descriptive statistics about the values in that column.

In [ ]:
# set float display format to comma as thousand separator and 2 digits after the decimal point
pd.options.display.float_format = '{:,.2f}'.format
demo_df['age'].describe()

,age
count,534.00
mean,21.40
std,2.98
min,18.00
25%,19.00
50%,20.00
75%,22.00
max,35.00


Here are something we can learn from these descriptive statistics about the students who completed the survey.

*  The average age for these students was 21.40.
*  The youngest student was 18.
*  The oldest student was 35.
*  A quarter of students were 19 or younger
*  Half of students were 20 or younger.
*  A quarter of students were 22 or above.

Is this a good representation of our class?

## Practice

In [ ]:
# Call the describe function to find the descriptive statistics for the 'units' and 'gpa' column.


**Add a text block** to explain what the above result tell you about the students who completed the survey.

# The unique and value_counts functions

To explore a column before cleaning it, pandas provides two helpful functions:
*  The unique() method shows all distinct values in a column
*  The value_counts() displays the frequency of each value.

In [ ]:
demo_df['gender'].unique()

array([1, 2, 3])

The above result shows that the 'gender' column contains three unique values: 1, 2, and 3.

In [ ]:
demo_df['gender'].value_counts()

,count
gender,
2,363
1,176
3,4


The above result shows that 362 students responded 2 (Female), 176 students responded 1 (Male), and 4 students responded 3 (Other).

## Practice

Use the unique and value_counts functions to explore categorical columns of your choice. You may use as many or as few code blocks as you'd like.

In [ ]:
# data exploration


**Add a text block** to share your findings about the categorical columns you chose to explore with the above code block(s).

# The replace function

The replace() function in pandas is a flexible tool for recoding values in a DataFrame. It allows you to substitute old values with new ones by passing in a dictionary.

The following code block replaces 1 in demo_df['children'] with "Yes" and 2 with "No". It is important to remember that the replace function creates a new column without changing the existing one.

In [ ]:
map = {1:'Yes', 2:'No'}
demo_df['children_text'] = demo_df['children'].replace(map)
demo_df[['children', 'children_text']].tail(10)

,children,children_text
533,2.0,No
534,1.0,Yes
535,2.0,No
536,2.0,No
537,2.0,No
538,2.0,No
539,2.0,No
540,2.0,No
541,2.0,No
542,2.0,No


If we want to update the children column instead of creating a new children_text column, the above assignment should have demo_df['children'] on the left of =.

## Practice

In [ ]:
# Use the replace function to change the gender column from 1, 2, 3 into Male, Female, and Other
# You may assign the result of the replace function back to the gender column or make a new gender_text column
